### PR TITLE
Component-based template validation PoC 

### PR DESCRIPTION
**Pull request by vlada-shubina**
### Problem
https://github.com/dotnet/templating/issues/2623

### Solution
Component-based template validation.
Note: the design done under assumption that `ITemplateValidator` and `ITemplateValidatorFactory` becomes public later on and any host or tool can create more validator components and inject the checks.
Note: current validation checks are only those which existed before, and serve as example. More checks to be added later on.

TODO: 
- [x] extract localization strings (once the approach is approved)

### Checks:
- [x] Added unit tests
- [x] Added `#nullable enable` to all the modified files [?](https://github.com/dotnet/templating/wiki/Contributing#coding-style)

### CODE REVIEW

File: src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/RunnableProjectGenerator.cs [hunk 1]
```diff
@@ -233,8 +236,10 @@ IParameterSet IGenerator.GetParametersForTemplate(IEngineEnvironmentSettings env
         [Obsolete]
         IList<ITemplate> IGenerator.GetTemplatesAndLangpacksFromDir(IMountPoint source, out IList<ILocalizationLocator> localizations)
         {
-            IReadOnlyList<ScannedTemplateInfo> foundTemplates = Task.Run(async () => await GetTemplatesFromMountPointInternalAsync(source, default).ConfigureAwait(false)).GetAwaiter().GetResult();
-            localizations = foundTemplates.SelectMany(t => t.Localizations.Values).ToList();
+            IReadOnlyList<ScannedTemplateInfo> foundTemplates =
+                      Task.Run(async () => await GetTemplatesFromMountPointInternalAsync(source, default).ConfigureAwait(false)).GetAwaiter().GetResult();
```
Comment (JanKrivanek) [inline src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/RunnableProjectGenerator.cs:240]: https://github.com/dotnet/templating/pull/5840/files#r1116713952

File: test/Microsoft.TemplateEngine.Orchestrator.RunnableProjects.UnitTests/SchemaTests/JSONSchemaTests.cs [hunk 2]
```diff
@@ -44,8 +44,12 @@ public void IsJSONSchemaValid(string testFile)
 
         public static IEnumerable<object?[]> GetAllTemplates()
```
Comment (JanKrivanek) [inline test/Microsoft.TemplateEngine.Orchestrator.RunnableProjects.UnitTests/SchemaTests/JSONSchemaTests.cs:45]: Should we run thouse through `ValidateAsync` as well (skipping the localization checking)?

**Review by JanKrivanek — APPROVED** [review]
This is great!

File: src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/Validation/ITemplateValidatorFactory.cs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+using System.Threading.Tasks;
+using Microsoft.TemplateEngine.Abstractions;
+
+namespace Microsoft.TemplateEngine.Orchestrator.RunnableProjects.Validation
+{
+    [Flags]
+    internal enum ValidationScope
+    {
+        None = 0,
+        Scanning = 1,
```
Comment (YuliiaKovalova) [inline src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/Validation/ITemplateValidatorFactory.cs:15]: is "Scanning" for installed templates?
what does it mean?

File: src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/Validation/MandatoryLocalizationValidationFactory.cs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+                                unusedManualInstructionLocs--;
+                            }
+                        }
+
+                        if (unusedManualInstructionLocs > 0)
+                        {
+                            // Localizations provide more translations than the number of manual instructions we have.
+                            string excessInstructionLocalizationIds = string.Join(
+                                ", ",
+                                postActionLocModel.Instructions.Keys.Where(k => !postAction.ManualInstructionInfo.Any(i => i.Id == k)));
```
Comment (YuliiaKovalova) [inline src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/Validation/MandatoryLocalizationValidationFactory.cs:72]: is it possible that we have a lot of concatenations here?

File: test/Microsoft.TemplateEngine.Orchestrator.RunnableProjects.UnitTests/SchemaTests/JSONSchemaTests.cs (hunk 2 above)
Comment (vlada-shubina) [reply test/Microsoft.TemplateEngine.Orchestrator.RunnableProjects.UnitTests/SchemaTests/JSONSchemaTests.cs:45]: added, nice idea! once rebased, we should also add tests for samples.

File: src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/Validation/MandatoryLocalizationValidationFactory.cs (hunk 4 above)
Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/Validation/MandatoryLocalizationValidationFactory.cs:72]: Not sure. This is existing code (now in https://github.com/dotnet/templating/blob/4fe046dbc4f7d079892750fa1b896c5a6356cd6e/src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/RunnableProjectConfig.cs#L306) and let's keep it as is for now.

**Review by YuliiaKovalova — APPROVED** [review]

**Comment by YuliiaKovalova** [comment]
looks good to me !

**Pull request merged**